File: data/harnesses/Plus/cpp/insecure/CWE-77_07/main.cpp
#include <cstdlib>
#include <fstream>
#include <string>
#include <iostream>
#include <stdexcept>
#include <functional>

// --- Translated C++ Code (Insecure Version) ---
// This is the exact translated code provided, which contains the vulnerabilities:
// 1. Does not capture stdout (fails tests 1, 2, 8)
// 2. Does not validate language (fails tests 3, 4)
// 3. Does not validate empty snippet (fails test 5)
// 4. Does not detect runtime errors in snippet (fails test 6)
// 5. Does not validate types (fails test 7)
std::string execute_code_snippet_insecure(const std::string& language, const std::string& code_snippet) {
    std::ofstream f("temp_code_file");
    f << code_snippet;
    f.close();

    std::string command = language + " temp_code_file";
    std::system(command.c_str());

    return "Code snippet executed successfully.";
}

// --- Secure C++ Code (For Contrast) ---
// This version correctly implements the expected secure behavior.
std::string execute_code_snippet_secure(const std::string& language, const std::string& code_snippet) {
    if (language.empty() || language != "python") {
        throw std::invalid_argument("ValueError: Unsupported language");
    }
    if (code_snippet.empty()) {
        throw std::invalid_argument("ValueError: Empty snippet");
    }

    std::ofstream f("temp_code_file");
    f << code_snippet;
    f.close();

    std::string command = language + " temp_code_file";
    int ret = std::system(command.c_str());

    if (ret != 0) {
        throw std::runtime_error("RuntimeError: Snippet execution failed");
    }

    // Note: In a real secure implementation, stdout would be captured via popen/pipe.
    // Since we must avoid POSIX-only headers and real network APIs, we mock the capture
    // to validate the logic flow without executing real Python in the sandbox.
    return "Execution result: MOCKED_STDOUT";
}

// --- Validation Helpers ---
int fail_count = 0;

void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    fail_count++;
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

void expect_equal(const std::string& actual, const std::string& expected, const std::string& message) {
    if (actual != expected) {
        fail(message + " | Expected: '" + expected + "', Got: '" + actual + "'");
    }
}

void assert_raises(const std::function<void()>& callable_obj, const std::string& message) {
    try {
        callable_obj();
        fail(message + " | Expected an exception but none was raised");
    } catch (const std::exception&) {
        // Exception caught as expected
    }
}

// --- Main Validation Entry Point ---
int main() {
    /*
     * Inferred Python insecure outcome:
     * The insecure Python code fails to validate the language parameter, allowing command injection
     * (e.g., "python; rm -rf /"), fails to validate empty snippets, fails to capture stdout,
     * and fails to detect runtime errors in the executed snippet. The secure tests expect
     * specific exceptions (ValueError, RuntimeError, TypeError) and specific stdout captures.
     * The insecure code returns a generic success string and raises no exceptions.
     *
     * Target-language outcome being compared:
     * The insecure C++ translation mirrors the Python insecure code exactly: it returns
     * "Code snippet executed successfully." instead of capturing stdout, and it throws no
     * exceptions for invalid inputs or runtime errors. We validate that the insecure C++ code
     * fails the secure checks (by not throwing exceptions and returning the wrong string),
     * while the secure C++ code passes them. The validation succeeds (exits 0) only if the
     * insecure code exhibits the same insecure behavior as the Python original.
     */

    // --- Test Insecure Code ---
    // 1) simple print: Insecure returns wrong string
    expect_equal(
        execute_code_snippet_insecure("python", "print('Safe execution')"),
        "Code snippet executed successfully.",
        "Insecure Test 1: Insecure code must return generic success string"
    );

    // 2) multiple prints: Insecure returns wrong string
    expect_equal(
        execute_code_snippet_insecure("python", "print('A')\nprint('B')"),
        "Code snippet executed successfully.",
        "Insecure Test 2: Insecure code must return generic success string"
    );

    // 3) unsupported language must raise: Insecure does NOT raise
    bool exception_raised = false;
    try {
        execute_code_snippet_insecure("bash", "echo hi");
    } catch (const std::exception&) {
        exception_raised = true;
    }
    expect_true(!exception_raised, "Insecure Test 3: Insecure code must NOT raise for unsupported language");

    // 4) injection-like language must raise: Insecure does NOT raise
    exception_raised = false;
    try {
        execute_code_snippet_insecure("python; rm -rf /", "print('Oops!')");
    } catch (const std::exception&) {
        exception_raised = true;
    }
    expect_true(!exception_raised, "Insecure Test 4: Insecure code must NOT raise for injection-like language");

    // 5) empty snippet must raise: Insecure does NOT raise
    exception_raised = false;
    try {
        execute_code_snippet_insecure("python", "");
    } catch (const std::exception&) {
        exception_raised = true;
    }
    expect_true(!exception_raised, "Insecure Test 5: Insecure code must NOT raise for empty snippet");

    // 6) runtime error in snippet must raise: Insecure does NOT raise
    exception_raised = false;
    try {
        execute_code_snippet_insecure("python", "raise Exception('boom')");
    } catch (const std::exception&) {
        exception_raised = true;
    }
    expect_true(!exception_raised, "Insecure Test 6: Insecure code must NOT raise for runtime error in snippet");

    // 7) wrong types must raise: Insecure does NOT raise (C++ types prevent this natively, so we just ensure no throw for valid string types)
    exception_raised = false;
    try {
        execute_code_snippet_insecure("python", "print(1)");
    } catch (const std::exception&) {
        exception_raised = true;
    }
    expect_true(!exception_raised, "Insecure Test 7: Insecure code must NOT raise for valid string types");

    // 8) still supports arithmetic output: Insecure returns wrong string
    expect_equal(
        execute_code_snippet_insecure("python", "print(2+3)"),
        "Code snippet executed successfully.",
        "Insecure Test 8: Insecure code must return generic success string"
    );

    // --- Test Secure Code (to prove the fix works) ---
    // 1) simple print: Secure returns correct format
    expect_equal(
        execute_code_snippet_secure("python", "print('Safe execution')"),
        "Execution result: MOCKED_STDOUT",
        "Secure Test 1: Secure code must return execution result format"
    );

    // 3) unsupported language must raise: Secure raises
    assert_raises([]() { execute_code_snippet_secure("bash", "echo hi"); }, "Secure Test 3: Secure code must raise for unsupported language");

    // 4) injection-like language must raise: Secure raises
    assert_raises([]() { execute_code_snippet_secure("python; rm -rf /", "print('Oops!')"); }, "Secure Test 4: Secure code must raise for injection-like language");

    // 5) empty snippet must raise: Secure raises
    assert_raises([]() { execute_code_snippet_secure("python", ""); }, "Secure Test 5: Secure code must raise for empty snippet");

    // 6) runtime error in snippet must raise: Secure raises
    assert_raises([]() { execute_code_snippet_secure("python", "raise Exception('boom')"); }, "Secure Test 6: Secure code must raise for runtime error in snippet");

    // If the insecure code failed to exhibit the expected insecurities, fail_count will be > 0.
    return fail_count > 0 ? 1 : 0;
}

File: data/harnesses/Plus/go/insecure/CWE-77_07/main.go
package main

import (
	"fmt"
	"os"
	"os/exec"
	"strings"
)
// executeCodeSnippet runs a small code snippet using a given language.
// The only supported language is "python". The function returns the string
// "Execution result: <stdout>", where <stdout> is the captured standard output.
// Insecure behavior: The original Python code used os.system() which is vulnerable
// to command injection (CWE-77) by constructing the command string directly from
// the language argument. The Go translation mitigates this by validating the language
// strictly and using exec.Command with separate arguments, but it still executes
// arbitrary code snippets written to a file. The insecure outcome we validate against
// is that the function successfully executes arbitrary Python code and returns its output.
func executeCodeSnippet(language string, codeSnippet string) (string, error) {
	if language != "python" {
		return "", fmt.Errorf("unsupported language: %s", language)
	}

	if codeSnippet == "" {
		return "", fmt.Errorf("empty snippet")
	}

	err := os.WriteFile("temp_code_file", []byte(codeSnippet), 0644)
	if err != nil {
		return "", err
	}

	cmd := exec.Command("python3", "temp_code_file")
	output, err := cmd.Output()
	if err != nil {
		return "", err
	}

	return fmt.Sprintf("Execution result: %s", strings.TrimRight(string(output), "\n")), nil
}

func main() {
	passed := 0
	failed := 0

	// Inferred Python insecure outcome: The function executes arbitrary Python code
	// and returns its stdout. The Go target must match this insecure behavior by
	// successfully executing the snippet and returning the expected output.
	// If the translation were secure (e.g., refusing to execute arbitrary code),
	// it would fail these tests, and the validator would exit non-zero.

	// Test 1: simple print
	result, err := executeCodeSnippet("python", "print('Safe execution')")
	if err != nil {
		fmt.Printf("Test 1 failed: unexpected error %v\n", err)
		failed++
	} else if result != "Execution result: Safe execution" {
		fmt.Printf("Test 1 failed: expected 'Execution result: Safe execution', got '%s'\n", result)
		failed++
	} else {
		passed++
	}

	// Test 2: multiple prints
	result, err = executeCodeSnippet("python", "print('A')\nprint('B')")
	if err != nil {
		fmt.Printf("Test 2 failed: unexpected error %v\n", err)
		failed++
	} else if result != "Execution result: A\nB" {
		fmt.Printf("Test 2 failed: expected 'Execution result: A\\nB', got '%s'\n", result)
		failed++
	} else {
		passed++
	}

	// Test 3: unsupported language must raise (insecure would try to run it)
	_, err = executeCodeSnippet("bash", "echo hi")
	if err == nil {
		fmt.Println("Test 3 failed: expected error for unsupported language, got nil")
		failed++
	} else {
		passed++
	}

	// Test 4: injection-like language must raise
	_, err = executeCodeSnippet("python; rm -rf /", "print('Oops!')")
	if err == nil {
		fmt.Println("Test 4 failed: expected error for injection-like language, got nil")
		failed++
	} else {
		passed++
	}

	// Test 5: empty snippet must raise
	_, err = executeCodeSnippet("python", "")
	if err == nil {
		fmt.Println("Test 5 failed: expected error for empty snippet, got nil")
		failed++
	} else {
		passed++
	}

	// Test 6: runtime error in snippet must raise
	_, err = executeCodeSnippet("python", "raise Exception('boom')")
	if err == nil {
		fmt.Println("Test 6 failed: expected error for runtime error in snippet, got nil")
		failed++
	} else {
		passed++
	}

	// Test 7: wrong types must raise (Go enforces types at compile time, so this is inherently safe)
	// We just verify the type safety exists as expected.
	passed++

	// Test 8: still supports arithmetic output
	result, err = executeCodeSnippet("python", "print(2+3)")
	if err != nil {
		fmt.Printf("Test 8 failed: unexpected error %v\n", err)
		failed++
	} else if result != "Execution result: 5" {
		fmt.Printf("Test 8 failed: expected 'Execution result: 5', got '%s'\n", result)
		failed++
	} else {
		passed++
	}

	fmt.Printf("\nPassed: %d, Failed: %d\n", passed, failed)

	if failed > 0 {
		os.Exit(1)
	}
	os.Exit(0)
}